components/builder-api/src/server/services/memcache.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2018 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use config::MemcacheCfg;
+use hab_core::package::PackageIdent;
+use memcache;
+use protobuf;
+use protobuf::Message;
+use protocol::originsrv::OriginPackage;
+
+pub struct MemcacheClient {
+    cli: memcache::Client,
+}
+
+impl MemcacheClient {
+    pub fn new(config: Vec<MemcacheCfg>) -> Self {
+        let memcache_host_strings: Vec<String> = config.iter().map(|h| format!("{}", h)).collect();
+        let memcache_hosts: Vec<&str> = memcache_host_strings.iter().map(AsRef::as_ref).collect();
+        MemcacheClient {
+            cli: memcache::Client::new(memcache_hosts).unwrap(),
+        }
+    }
+    // TED: we'll probably want a more generic set/get methods that are strings only
+    pub fn set_package(&mut self, package: OriginPackage, channel: &str) {
+        match self.cli.set(
+            &format!("{}/{}", channel, package.to_string()),
+            package.write_to_bytes().unwrap().as_slice(),
+            15 * 60,
+        ) {
+            Ok(_) => debug!("Saved {:?} to memcached", package.to_string()),
+            Err(e) => warn!(
+                "Failed to save {:?} to memcached: {}",
+                package.to_string(),
+                e
+            ),
+        };
+    }
+    pub fn get_package(&mut self, package: PackageIdent, channel: &str) -> Option<OriginPackage> {
+        match self.get(&format!("{}/{}", channel, package.to_string())) {
+            Some(ops) => Some(protobuf::parse_from_bytes(&ops).unwrap()),
+            None => None,
+        }
+    }
+    pub fn delete_package(&mut self, package: PackageIdent, channel: &str) {
+        match self
+            .cli
+            .delete(&format!("{}/{}", channel, package.to_string()))
+        {
+            Ok(_) => debug!("Deleted {:?} from memcached", package.to_string()),
+            Err(e) => warn!(
+                "Failed to delete {:?} from memcached: {}",
+                package.to_string(),
+                e
+            ),
+        }
+    }
+
+    fn get(&mut self, key: &str) -> Option<Vec<u8>> {
```

Comment:
does this get used?

Reply:
It does. In the package_get function. I have plans for it to be used for session tokens and I didn't want to duplicate the cli calls. Since it's a 1-liner now it's 6 of one, half dozen the other